File: Data Structures & Algorithms/perfect-squares/submission-0.py
class Solution:
    def numSquares(self, n: int) -> int:
        # find smallest sum of perfect squares to n
        # dp[1...n] where dp[i] is smallest sum of perfect square to add to i
        # dp[i] = for p in perfectsqaures: dp[i] = min(dp[i-p]+1,dp[i])
        dp = [10000]*(n+1)
        dp[0] = 0
        dp[1] = 1
        for i in range(1,n+1):
            m = math.floor(math.sqrt(i)) #largest possible sqrt possible for this index
            print(i,m)
            for j in range(1,m+1):
                dp[i] = min(dp[i-(j**2)]+1,dp[i])
        print(dp)
        return dp[-1]

File: Data Structures & Algorithms/max-water-container/submission-0.java
class Solution {
    public int maxArea(int[] heights) {
        int maxArea = 0;
        int p1 = 0;
        int p2 = heights.length - 1;
        while (p1 <= heights.length-1 && p2 >= 0){
            int curArea = Math.abs((p1-p2))*(Math.min(heights[p2],heights[p1]));
            maxArea = (curArea>maxArea) ? curArea : maxArea;
            if (heights[p1]>heights[p2]){
                p2--;
            }else{
                p1++;
            }
            // else:p1 ++
            
        }
        return maxArea;
    }
}
